### DISCUSSION TITLE
APIs about achieving "Reader Mode" which is available in IntelliJ

### DISCUSSION THREAD

**Discussion by mossgowild** [microsoft/vscode-discussions#432] [Extension Development QnA] [answered]
I plan to write an extension to achieve the "Reader Mode" as IntelliJ does. According to the vscode API document, there may be a lack of APIs to implement the functionalities below:
1. Comments in block ghost text.
  1.1 The font family and font size are different from the editor font.
  1.2 Comment text is retrieved from language servers.
  1.3 Hiding original comment text from code.
2. File is read-only and has semantic highlight and language server support.

I tried Virtual Document API to implement read-only files but it lost the support of language servers.
Are there some APIs I've missed? Or will vscode add these APIs in the future?

**Comment by r3m0t** [comment, marked as answer]
You can use the request forwarding  design pattern to get back semantic highlight support.

Call `registerDocumentSemanticTokensProvider` to register the feature for your URI scheme, and implement `provideDocumentSemanticTokens` as calling the built in command `vscode.provideDocumentSemanticTokens` with the original URI and returning the result.

The other language features can be implemented similarly.

**Reply by mossgowild** [reply]
Thank you, I'll try.

**Reply by mossgowild** [reply]
Thanks for your suggestions, almost all language features are hijacked, except that `vscode.provideDocumentSemanticTokens` doesn't provide comment tokens, any solution for this?

**Reply by mossgowild** [reply]
I see other extension authors using textmate or regex to get comment position, maybe this's the only solution because vscode doesn't provide this built-in command.

**Reply by r3m0t** [reply]
The semantic tokens is only information that layers on top of the textmate language. I don't think there's any way to access the same grammars or parser that vs code is using, but you can probably find it on npm.

**Reply by mossgowild** [reply]
Thanks, it was finally done with regular expression and `vscode.FileSystemProvider`.

**Comment by mossgowild** [comment]
The extension has been published now, see [Reader Mode](https://marketplace.visualstudio.com/items?itemName=theoolee.reader-mode).
